### ISSUE TITLE
Serial data between the stick and openzwave

### ISSUE THREAD

**Issue by kdschlosser** [OpenZWave/python-openzwave#185]
This is a request and not an issue. I am going to write a new unittest for OpenZWave . the current one does not test functionality of the library at all. I would like to test all of the functionality of the library and in order to do that I need to know what the data is that is being passed between the adapter and OpenZWave .

My idea is to create a pair of virtual serial ports and "wire" them together like a null modem cable would. this way I can connect to one port in the unittests and have OpenZWave connect to the other... This is easy enough to do on Windows, OSX and Linux.

So this is what I would like to have..

a pseudo network that has devices in it that cover all of the command classes that OpenZWave supports. I would need to have the commands that are sent by OpenZWave and the expected responses from the stick. It would be great to have them for any/all commands that the nodes support.

I know this is a tall order I didn't know if there was a way to spit this information out of OpenZWave or not. I set up a monitor to watch the data flowing back and forth and I have tried to marry a command to a response.. I am not having any success doing this.

If this can be done that would be a HUGE help.

Thanks

**Comment by petergebruers** [comment]
I have written a few (incomplete) device simulators which do exactly that. I have not published them because they serve or served a specific purpose plus I am still trying to get to a more universal solution and find "the right way to do things"

Runs on linux and mac only, there are timers (which can be replaced with c++11 or o/s specific stuff) and on linux I use virtual tty to emulate the com port.

Basically it implements a controller and feeds/analyses data like OpenZWave. The protocol is quite symmetrical. I get the right data either by constructing it or by extracting (bits) manually from an OZW log.

I'll contact you on discord or slack this weekend.

**Comment by kdschlosser** [comment]
Putting together an emulator is a snap in python. I have to do some more reading on setting up a null modem virtual serial port pair on Linux but from the information I have read it is really easy to do. There is com0com for Windows which has a command line utility for creating and managing virtual serial port pairs..

The absolute worst case scenario is I would use the remote connection code I wrote and have it connect to my test setup here at home. It's basically a wrapper around the manager and the notifications. But it would allow me to run tests on portions of the code. I do not have access to all of the various devices that would be needed to cover all of the supported command classes. Thatis why I would prefer the method i described above. I just need to know the commands and expected responses in order to be able to run the tests.

I wonder if this could be done... can a single pseudo device be constructed? one that would be able to support all of the command classes??
